src/client.rs

```
@@ -0,0 +1,82 @@
+use std::sync::Arc;
+
+use crate::{
+    concern::{ReadConcern, WriteConcern},
+    db::Database,
+    error::Result,
+    options::DatabaseOptions,
+    read_preference::ReadPreference,
+};
+
+/// This is the main entry point for the API. A `Client` is used to connect to a MongoDB cluster.
+/// By default, it will monitor the topology of the cluster, keeping track of any changes, such
+/// as servers being added or removed
+///
+/// `Client` uses [`std::sync::Arc`](https://doc.rust-lang.org/std/sync/struct.Arc.html) internally,
+/// so it can safely be shared across threads. For example:
+///
+/// ```
+/// 
+/// let client = Client::connect("mongodb://example.com");
+///
+/// for i in range 0..5 {
+///     let client_ref = client.clone();
+///
+///     std::thread::spawn(move || {
+///         let collection = client_ref.database("items").collection(format!("coll{}", i));
+///
+///         // Do something with the collection
+///     });
+/// }
+/// ```
+#[derive(Clone, Debug)]
+pub struct Client {
+    inner: Arc<ClientInner>,
+}
+
+#[derive(Debug)]
+struct ClientInner {
+    read_preference: Option<ReadPreference>,
+    read_concern: Option<ReadConcern>,
+    write_concern: Option<WriteConcern>,
+}
+
+impl Client {
+    /// Creates a new `Client` connected to the cluster specified by `uri`. `uri` must be a valid
+    /// MongoDB connection string.
+    pub fn connect(uri: &str) -> Result<Self> {
```

Does this actually result in a connection to a cluster? In most drivers (unfortunately node isn't quite there yet), I/O is deferred to the first operation. In node, we have a `MongoClient.connect` which has lead to a lot of confusion about what it even means to connect (am I connected when I connect to the primary? A primary and two secondaries? etc.). If no connection is actually made here, might we consider a different name for the method?

---

This is a really good point. The go driver's `Connect`, for example, does not actually attempt to connect to the server; you have to do a `Connect` and then a `Ping` (which I only found out recently, a number of tools tests were silently succeeding because of this).

---

In the current implementation, this method starts the monitoring threads for each of the hosts given in the connection string but does not do any blocking I/O. No additional methods need to be called to be able to perform operations; you can immediately create a client and then being using it. However, any errors due to being unable to reach a server won't occur until the first operation is attempted (which is my understanding of what the SDAM spec requires).

As for whether the client is connected to a primary or secondaries, it will be connected to exactly the list of hosts specified in the connection string until monitoring threads update the topology based on the isMaster responses. Given that the monitoring threads will execute isMasters immediately when started (not sleeping `heartbeatFrequencyMS` until the bottom of the loop), the first topology updates should easily happen before server selection times out, so there shouldn't be any issues with users being unable to perform operations requiring a certain server type immediately after creating the client as long as the driver is able to reach the cluster and the cluster contains a server of the required type.

I'm fine with changing the name of this method if you can think of something more descriptive; the main reason I called it this is that I wanted to convey that more stuff is happening under the hood that a user might expect from a typical `Foo::new` method in Rust, which often is just a thin wrapper around a struct literal with the arguments as the fields.

---

Ah, okay. The node variant of this method actually blocks until some odd sense "connection" occurs (it was originally when the primary was connected to, so not sufficient in the cases when you wanted to target e.g. a secondary only). What you are doing sounds like what most drivers do in their `MongoClient` constructor. Is it not idiomatic to put this into some sort of `Client::new` method instead?

---

We could do `new` (notwithstanding the rationale I gave above for not using it), but given that we plan to add another constructor for using options, it's probably better to just make two methods `Client::with_uri` and `Client::with_options` (which is the naming idiom for constructors which take specific parameters). Assuming there are no objections to this, I'll rename this method (but will hold off on adding the `with_options` API for the same reason I gave in my top-level response to Patrick.

---

One final question I have with regards to this: do we want to define a uri type similar to `mongoc_uri_t` that we would take here instead of a string?

---

I'm not convinced that it's necessary, but the implementation does already have a type like that that it uses internally, so it would be super easy to just expose that if we decide to go that route.

---

Actually, now that I think of it, there's any easy way to both add the options API quickly and have a public-facing connection string type; since the internal connection string type is already a struct that contains all of the available options (along with a set of host/port pairs), we could just make the fields public and make a constructor that takes that as a parameter. `Client::with_uri` could be renamed `Client::with_uri_str`, and it would just call `Client::with_options` with the result of the parsed string.

---

so would the `with_options` function be taking in a uri type? Or would the uri type be publicly exported as something options-y that also has uri string parsing capabilities?

---

The latter; I'm not really sure what could even go into a URI type rather than a set of configuration options and the ability to parse it from a string, and calling it something URI-related doesn't seem like it will do anything besides making it less obvious that users can just create it and use it without needing to do anything string related.

---

Oh, and we could have a method that converts it back to a string in case users need that for some reason. Depending on how it was constructed, we could potentially even save the original URI it was constructed with if we wanted to preserve the ordering of the options.

---

That all sounds good to me!